### PR TITLE
feat(rpc-client): Introduce rpc-jsonrpsee Crate

### LINKED ISSUE #32: [Feature] Rpc type for `optimism_outputAtBlock`
### Component

rpc

### Describe the feature you would like

Add a type for the output of `optimism_outputAtBlock` endpoint.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by refcell**
**Description**

Introduces an `op-alloy-rpc-jsonrpsee` crate that can replace the non-type logic `reth/optimism/rpc`. Type logic can go in `op-alloy/rpc-types`.

Closes #32

### CODE REVIEW

**Comment by refcell** [comment]
> [!NOTE]
>
> Depends on #35

File: crates/rpc-client/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+# rpc
+jsonrpsee.workspace = true
+jsonrpsee-types.workspace = true
+
+[features]
+client = [
+    "jsonrpsee/client",
+    "jsonrpsee/async-client",
+]
```
Comment (mattsse) [inline crates/rpc-client/Cargo.toml:27]: hmm by default this now generates jsonrpsee server traits, so perhaps `rpc-client` isn't the perfect fit here.

**Review by mattsse — CHANGES_REQUESTED** [review]
perhaps we rename this to op-alloy-jsonrpsee ?

File: crates/rpc-jsonrpsee/src/traits.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    #[method(name = "syncStatus")]
+    async fn optimism_sync_status(&self) -> RpcResult<SyncStatus>;
+
+    /// Get the rollup configuration parameters.
+    #[method(name = "rollupConfig")]
+    async fn optimism_rollup_config(&self) -> RpcResult<RollupConfig>;
+
+    /// Get the software version.
+    #[method(name = "version")]
+    async fn optimism_version(&self) -> RpcResult<String>;
```
Comment (mattsse) [inline crates/rpc-jsonrpsee/src/traits.rs:30]: should we shorten all of these to `op_` and stick with that convention?

Comment (refcell) [reply crates/rpc-jsonrpsee/src/traits.rs:30]: Good callout - I like this a lot more

**Comment by refcell** [comment]
> [!NOTE]
>
> Adds the `OutputResponse` type to `rpc-types` since it's used by the trait. Takes care of #32

**Review by HrikB — COMMENTED** [review]
I wanna ask if it makes sense to implement a `Provider` here over the `Optimism` network. I was doing smth like this:

```
#[async_trait]
pub trait OpStackProvider<T: Transport + Clone = BoxTransport>: Provider<T, Optimism> {
    async fn output_at_block(&self, block_number: u64) -> TransportResult<OutputAtBlock> {
        let response = self
            .client()
            .request("optimism_outputAtBlock", (format!("0x{:x}", block_number),))
            .await?;

        Ok(response)
    }
}

impl<T: Transport + Clone> OpStackProvider<T> for RootProvider<T, Optimism> {
}
```

**Comment by mattsse** [comment]
yep this makes sense, mind opening an issue for this?
this should go into a new provider crate

**Review by mattsse — APPROVED** [review]

**Pull request merged**